selfdrive/boardd/boardd.cc

```diff
@@ -208,38 +208,31 @@ void can_send_thread(std::vector<Panda *> pandas, bool fake_send) {
   set_thread_name("boardd_can_send");
 
   AlignedBuffer aligned_buf;
-  Context * context = Context::create();
-  SubSocket * subscriber = SubSocket::create(context, "sendcan");
+  std::unique_ptr<Context> context(Context::create());
+  std::unique_ptr<SubSocket> subscriber(SubSocket::create(context.get(), "sendcan"));
   assert(subscriber != NULL);
   subscriber->setTimeout(100);
 
   // run as fast as messages come in
   while (!do_exit && check_all_connected(pandas)) {
-    Message * msg = subscriber->receive();
+    std::unique_ptr<Message> msg(subscriber->receive());
     if (!msg) {
       if (errno == EINTR) {
         do_exit = true;
       }
       continue;
     }
 
-    capnp::FlatArrayMessageReader cmsg(aligned_buf.align(msg));
+    capnp::FlatArrayMessageReader cmsg(aligned_buf.align(msg.get()));
     cereal::Event::Reader event = cmsg.getRoot<cereal::Event>();
 
     //Dont send if older than 1 second
-    if (nanos_since_boot() - event.getLogMonoTime() < 1e9) {
-      if (!fake_send) {
-        for (const auto& panda : pandas) {
-          panda->can_send(event.getSendcan());
-        }
+    if ((nanos_since_boot() - event.getLogMonoTime() < 1e9) && !fake_send) {
+      for (const auto& panda : pandas) {
+        panda->can_send(event.getSendcan());
       }
     }
-
-    delete msg;
   }
-
-  delete subscriber;
-  delete context;
```

Comment:
For ZMQ it's important to delete all subscribers before terminating the context, or that might hang forever. It seems like unique pointers are destructed in reverse order of their creation, so that should work out here. Is there any way we can make this relation more explicit?

Reply:
we can use braces to make it "more explicit",  But I think it might be redundant since objects are always destructed in the reverse order of construction (LIFO). 

```
unique_ptr ctx
{
  unique_ptr msg
}
```

Reply:
Guess it's fine since it's impossible to create the subscriber before the context.